gn: 2020-03-09 -> 2024-05-13

skeuchel opened the pull request:
###### Description of changes

This updates gn to a version which includes support for riscv64 builds. The revision used for gn is the one specified for the current chromium stable version (111.0.5563.64).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
    I've build the package and chromium which overrides this package on staging.
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

skeuchel commented:
To reproduce the build failure on the current master branch you can run
```
nix build nixpkgs#pkgsCross.riscv64.gn
```

JamieMagee commented on pkgs/development/tools/build-managers/gn/default.nix:
```diff
@@ -2,9 +2,9 @@
-  revNum = "1718"; # git describe HEAD --match initial-commit | cut -d- -f3
-  version = "2020-03-09";
-  sha256 = "1asc14y8by7qcn10vbk467hvx93s30pif8r0brissl0sihsaqazr";
+  # https://github.com/archlinux/svntogit-packages/blob/packages/gn/trunk/chromium-gn-version.sh
```
Arch Linux recently completed their git migration: https://archlinux.org/news/git-migration-completed/

```suggestion
  # https://gitlab.archlinux.org/archlinux/packaging/packages/gn/-/blob/main/chromium-gn-version.sh
```

skeuchel replied:
The (OmahaProxy)[https://omahaproxy.appspot.com] service which was used by that script is now permanently offline and Arch simply dropped the script for good. Instead I am referring to the chromium nix file, which contains the gn version/rev as well.

SeTSeR commented:
There is already gn of 2021-08-08 in nixpkgs, and this version already includes risc-v support. https://github.com/NixOS/nixpkgs/blob/b4fc1d4a6f6cfdebf711fd00be7a5f13ace2af1b/pkgs/development/tools/build-managers/gn/rev1924.nix

Bump of default gn version may introduce breakage in dependent packages, as gn provides no guarantees for backwards compatibility. Maybe just explicitly use `gn1924` for dependencies needed in RISC-V?

Nyabinary commented:
Any news on this happening?

SeTSeR commented:
> Any news on this happening?

There seem to be four packages, depending on `gn`, namely `skia`, `v8`, `oni2` and `chromium`:
https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+gn%2C&type=code

If this bump doesn’t break them, it should be safe to make it. However, I currently don’t know how to do it, I don’t have access to x86 machine with Nix.

SeTSeR commented:
@ofborg build envoy v8 oni chromium libsForQt5.qtwebengine libsForQt6.qtwebengine neovide python3Packages.skia-pathops aseprite

Nyabinary commented:
@SeTSeR Did ofbrog ever attempt to build?

SeTSeR commented:
> @SeTSeR Did ofbrog ever attempt to build?

I don't have much experience with ofborg myself, as I don't run it much, but it seems something was run as we can see from the checks:
https://logs.ofborg.org/?key=nixos/nixpkgs.221306&attempt_id=bff6c52c-bccf-4e1c-b620-ac54cd863f29
However, I can't access logs now, maybe the check is too old and I have to start it again.

SeTSeR commented:
@ofborg build envoy v8 oni chromium libsForQt5.qtwebengine libsForQt6.qtwebengine neovide python3Packages.skia-pathops aseprite

skeuchel commented:
I have rebased this PR and changed it to bump to the gn version (2024-01-22) used in the current chromium stable version. Only a handful of packages use gn, and some always override it. The **exhaustive list** is

#### non-override uses
- [aseprite](https://github.com/NixOS/nixpkgs/blob/a3e2b0de906a8fe0143c2783199abdc132dee56a/pkgs/applications/editors/aseprite/skia.nix#L22)
- [neovide](https://github.com/NixOS/nixpkgs/blob/a3e2b0de906a8fe0143c2783199abdc132dee56a/pkgs/applications/editors/neovim/neovide/default.nix#L61)
- [libsForQt5.qt5.qtwebengine](https://github.com/NixOS/nixpkgs/blob/a3e2b0de906a8fe0143c2783199abdc132dee56a/pkgs/development/libraries/qt-5/modules/qtwebengine.nix#L65)
- [qt6.qtwebengine](https://github.com/NixOS/nixpkgs/blob/a3e2b0de906a8fe0143c2783199abdc132dee56a/pkgs/development/libraries/qt-6/modules/qtwebengine.nix#L106)
- [python31XPackages.skia-pathops](https://github.com/NixOS/nixpkgs/blob/c5101e457206dd437330d283d6626944e28794b3/pkgs/development/python-modules/skia-pathops/default.nix#L31)

#### override uses
- [chromium](https://github.com/NixOS/nixpkgs/blob/a3e2b0de906a8fe0143c2783199abdc132dee56a/pkgs/applications/networking/browsers/chromium/default.nix#L56)
- [v8](https://github.com/NixOS/nixpkgs/blob/a3e2b0de906a8fe0143c2783199abdc132dee56a/pkgs/development/libraries/v8/default.nix#L67)

I have successfully built all of them on `x86_64-linux`. The package `oni2` was removed from nixpkgs since opening this PR.

skeuchel commented:
Pinging maintainers of the maintainers of the direct reverse deps: skia-pathops (@barinov274), qt5.qtwebengine (@matthewbauer) and of qt-6 modules (@milahu, @nickcao). Can you take a look at this as well?

tobim commented on pkgs/development/tools/build-managers/gn/default.nix:
```diff
@@ -1,10 +1,10 @@
-  version = "2020-03-09";
-  sha256 = "1asc14y8by7qcn10vbk467hvx93s30pif8r0brissl0sihsaqazr";
+  # Note: Please use the recommended version for Chromium stabe, i.e. from
+  # <nixpkgs>/pkgs/applications/networking/browsers/chromium/upstream-info.nix
+  rev = "f99e015ac35f689cfdbf46e4eb174e5d2da78d8e";
+  revNum = "2140"; # git describe $rev --match initial-commit | cut -d- -f3
+  version = "2024-01-22";
```
Can you synchronize this again?

skeuchel committed:
gn: 2020-03-09 -> 2024-05-13
